LOOP-1200: Dump Alert database into Issue Report

rickpasetto opened the pull request:
Added query by date, plus unit tests.

rickpasetto committed:
LOOP-1200: Dump Alert database into Issue Report

Added query by date, plus unit tests.

rickpasetto commented on Loop/Managers/Alerts/DeviceAlertManager.swift:
```diff
@@ -146,29 +152,7 @@ extension DeviceAlertManager {
     
 }
 
-extension FileManager {
```
This code moved, unchanged, below all of the `extension DeviceAlertManager` declarations

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -149,8 +149,8 @@ extension AlertStore {
-
-    func executeAlertQuery(fromQueryAnchor queryAnchor: QueryAnchor? = nil, limit: Int, completion: @escaping (QueryResult) -> Void) {
+    
+    func executeAlertQuery(from queryAnchor: QueryAnchor? = nil, since date: Date? = nil, limit: Int, completion: @escaping (QueryResult) -> Void) {
```
This function signature, as written, does not make sense to me. It implies you could specify both a query anchor *and* a date, but that makes no sense. 

If you want to have common code, then when I'd recommend is having a common `executeAlertQuery` private function that takes an `NSPredicate` argument. Then add two "covering" public or internal functions, one that takes a query anchor (and, please, match the other remote data services function signature for consistency) and another that takes a non-optional date.

Specifically:
```
func executeAlertQuery(fromQueryAnchor queryAnchor: QueryAnchor?, limit: Int, completion: @escaping (QueryResult) -> Void)
func executeAlertQuery(since date: Date, limit: Int, completion: @escaping (QueryResult) -> Void)
private func executeAlertQuery(with predicate: NSPredicate, limit: Int, completion: @escaping (QueryResult) -> Void)
```

(Removing the default `nil` queryAnchor also will help reduce confusion about which version to call.)

rickpasetto replied:
Yeah I see your point.  Let me take a crack at it.

darinkrauss commented on Loop/Managers/Alerts/DeviceAlertManager.swift:
```diff
@@ -246,6 +230,61 @@ extension DeviceAlertManager {
+extension DeviceAlertManager {
+    
+    func getStoredEntries(startDate: Date, completion: @escaping (_ report: String) -> Void) {
+        alertStore.executeAlertQuery(since: startDate, limit: 100) {
```
[nit] You've got multiple uses of `$0` in this block which lead to my confusion. I'd suggest using named variables, `result` and `entry` specifically.

darinkrauss commented on Loop/Managers/Alerts/DeviceAlertManager.swift:
```diff
@@ -246,6 +230,61 @@ extension DeviceAlertManager {
+                    return """
+                    **\($0.title ?? "??")**
+                    
+                    * identifier: \($0.identifier!)
```
[nit] Going back to the model, I guess, if `identifier` and `issuedDate` are non-optional, why not define them that way?

rickpasetto replied:
`StoredAlert`'s members are auto-generated.   I don't know why they are not non-Optional in Swift :(.  I could manually write it, but I'd prefer not to, if possible.

darinkrauss replied:
One thing to note is all of the other Core Data models in LoopKit are hand-written. Checkout all of the `*+CoreDataClass.swift` and `*+CoreDataProperties.swift` files.

rickpasetto replied:
Thanks!  I saw that.  I'm not a huge fan of maintaining (and writing unit tests for) boilerplate code that Xcode can generate for us.  The downside is weird stuff like this, not obeying the "Optional" declaration in the model file.  It felt to me minor enough to not warrant manually writing it.  If you feel strongly, I'm happy to do it though.

darinkrauss replied:
Up to you. 

I don't want to get off into the weeds, but one of my developer idiosyncrasies is my preference for consistency above most else (well, other than correctness, of course).

Having multiple different implementations doing effectively the same thing is just harder to grok over time with large codebases. If fact, in the code I'm working on right now I just found three custom implementations of "purgeObjects". Just replaced them with the common implementation used by the rest of the code.

darinkrauss commented on Loop/Managers/DeviceDataManager.swift:
```diff
@@ -252,46 +252,51 @@ final class DeviceDataManager {
-                    loopReport,
-                ].joined(separator: "\n")
+            
+            self.deviceAlertManager?.getStoredEntries(startDate: Date() - .hours(48)) { (alertReport) in
```
[nit] If `deviceAlertManager` is `nil` then none of the rest will be executed (including the deviceLog and the rest).  Is `deviceAlertManager` really optional?

rickpasetto replied:
It's a weak var, initialized in `init`. If it is `nil`, it is a coding error. I suppose I could force-unwrap here to make that clear...or declare the weak var with `!`.  Either way this would be fatal.
WDYT?

darinkrauss replied:
That's up to you. I was just pointing it out. I have some low level of concern for these types of things because I consider them hidden issues. If, somehow, `deviceAlertManager` was nil under certain circumstances, it is highly unlikely we'd find it since, at least in this case, is only visible if you just happen to notice it missing in a diagnostic report.

darinkrauss commented on Loop/Managers/Alerts/DeviceAlertManager.swift:
```diff
@@ -246,6 +230,61 @@ extension DeviceAlertManager {
+    func getStoredEntries(startDate: Date, completion: @escaping (_ report: String) -> Void) {
+        alertStore.executeAlertQuery(since: startDate, limit: 100) {
+            switch $0 {
+            case .failure(let error): completion("Error: \(error)")
```
[nit] I know this saves a whole line of code 😉, but honestly I just completely tend to miss the statement after the case when reading the code. My preference (take with a grain of salt) is to have the `completion` call on a separate line.

I've never been a fan of a single line that includes both conditional and statement (e.g. `if x { y }` or `guard x else { y }`), but even less so when it is embedded in a large block of code.

My $0.02.

rickpasetto replied:
👍 In this case, good point, you are right: this should be on a separate line, but (again, IMHO) because the other `case` is not a single line.  When all of the `case` statements are short, I very much prefer this:
```swift
switch number {
    case .one: return 1
    case .two: return 2
    case .three: return 3
}
```
over this:
```swift
switch number {
    case .one: 
        return 1
    case .two: 
        return 2
    case .three: 
        return 3
}
```
Again, just MHO.

darinkrauss reviewed (changes requested):
I think the `executeAlertQuery` function needs to be updated per my comment. If you feel strongly about the current implementation, then I'll approve, but would prefer the change. Otherwise, LGTM. 📟

rickpasetto committed:
PR Feedback, plus some rework/refactor

This takes @darin's suggestion, with some modifications :).  Now, the Filter is part of the Anchor, which makes it so anchoring continues to apply the filter as appropriate.

darinkrauss reviewed (approved):
LGTM.
